File: manitas-creativas-frontend/src/services/rubroUniformeDetalleService.ts
import { makeApiRequest } from './apiHelper';

// Types for RubroUniformeDetalle
export interface RubroUniformeDetalle {
  id: number;
  rubroId: number;
  prendaUniformeId: number;
  rubroDescripcion: string;
  prendaUniformeDescripcion: string;
  prendaUniformeSexo: string;
  prendaUniformeTalla: string;
  prendaUniformePrecio: number;
  fechaCreacion: string;
  fechaActualizacion?: string;
  usuarioCreacionId: number;
  usuarioActualizacionId?: number;
  esEliminado: boolean;
  motivoEliminacion?: string;
  fechaEliminacion?: string;
  usuarioEliminacionId?: number;
}

export interface RubroUniformeDetalleCreate {
  rubroId: number;
  prendaUniformeId: number;
}

class RubroUniformeDetalleService {
  private readonly baseUrl = '/rubro-uniforme-detalles';

  // Get all rubro uniforme detalles
  async getAllRubroUniformeDetalles(): Promise<RubroUniformeDetalle[]> {
    return await makeApiRequest<RubroUniformeDetalle[]>(`${this.baseUrl}`, 'GET');
  }

  // Get active rubro uniforme detalles
  async getActiveRubroUniformeDetalles(): Promise<RubroUniformeDetalle[]> {
    return await makeApiRequest<RubroUniformeDetalle[]>(`${this.baseUrl}/active`, 'GET');
  }

  // Get rubro uniforme detalle by ID
  async getRubroUniformeDetalleById(id: number): Promise<RubroUniformeDetalle> {
    return await makeApiRequest<RubroUniformeDetalle>(`${this.baseUrl}/${id}`, 'GET');
  }

  // Get rubro uniforme detalles by rubro ID
  async getRubroUniformeDetallesByRubroId(rubroId: number): Promise<RubroUniformeDetalle[]> {
    return await makeApiRequest<RubroUniformeDetalle[]>(`${this.baseUrl}/by-rubro/${rubroId}`, 'GET');
  }

  // Get rubro uniforme detalles by prenda uniforme ID
  async getRubroUniformeDetallesByPrendaId(prendaUniformeId: number): Promise<RubroUniformeDetalle[]> {
    return await makeApiRequest<RubroUniformeDetalle[]>(`${this.baseUrl}/by-prenda/${prendaUniformeId}`, 'GET');
  }

  // Get specific rubro uniforme detalle by rubro and prenda
  async getRubroUniformeDetalleByRubroAndPrenda(rubroId: number, prendaUniformeId: number): Promise<RubroUniformeDetalle> {
    return await makeApiRequest<RubroUniformeDetalle>(`${this.baseUrl}/by-rubro-prenda/${rubroId}/${prendaUniformeId}`, 'GET');
  }

  // Create new rubro uniforme detalle
  async createRubroUniformeDetalle(data: RubroUniformeDetalleCreate, usuarioCreacionId: number): Promise<RubroUniformeDetalle> {
    return await makeApiRequest<RubroUniformeDetalle>(`${this.baseUrl}?usuarioCreacionId=${usuarioCreacionId}`, 'POST', data);
  }

  // Update rubro uniforme detalle
  async updateRubroUniformeDetalle(id: number, data: RubroUniformeDetalleCreate, usuarioActualizacionId: number): Promise<RubroUniformeDetalle> {
    return await makeApiRequest<RubroUniformeDetalle>(`${this.baseUrl}/${id}?usuarioActualizacionId=${usuarioActualizacionId}`, 'PUT', data);
  }

  // Delete rubro uniforme detalle (soft delete)
  async deleteRubroUniformeDetalle(id: number, motivoEliminacion: string, usuarioEliminacionId: number): Promise<void> {
    const params = new URLSearchParams({
      motivoEliminacion,
      usuarioEliminacionId: usuarioEliminacionId.toString(),
    });

    await makeApiRequest<void>(`${this.baseUrl}/${id}?${params}`, 'DELETE');
  }

  // Check if combination exists
  async existsRubroUniformeDetalle(rubroId: number, prendaUniformeId: number): Promise<boolean> {
    const response = await makeApiRequest<{ exists: boolean }>(`${this.baseUrl}/exists/${rubroId}/${prendaUniformeId}`, 'GET');
    return response.exists;
  }
}

export const rubroUniformeDetalleService = new RubroUniformeDetalleService();

File: ManitasCreativas.Application/Services/RubroUniformeDetalleService.cs
using ManitasCreativas.Application.DTOs;
using ManitasCreativas.Application.Interfaces.Repositories;
using ManitasCreativas.Application.Interfaces.Services;
using ManitasCreativas.Domain.Entities;

namespace ManitasCreativas.Application.Services;

public class RubroUniformeDetalleService : IRubroUniformeDetalleService
{
    private readonly IRubroUniformeDetalleRepository _rubroUniformeDetalleRepository;
    private readonly IRubroRepository _rubroRepository;
    private readonly IPrendaUniformeRepository _prendaUniformeRepository;
    private readonly IUsuarioRepository _usuarioRepository;

    public RubroUniformeDetalleService(
        IRubroUniformeDetalleRepository rubroUniformeDetalleRepository,
        IRubroRepository rubroRepository,
        IPrendaUniformeRepository prendaUniformeRepository,
        IUsuarioRepository usuarioRepository)
    {
        _rubroUniformeDetalleRepository = rubroUniformeDetalleRepository;
        _rubroRepository = rubroRepository;
        _prendaUniformeRepository = prendaUniformeRepository;
        _usuarioRepository = usuarioRepository;
    }

    public async Task<IEnumerable<RubroUniformeDetalleDto>> GetAllAsync()
    {
        var detalles = await _rubroUniformeDetalleRepository.GetAllWithRelatedDataAsync();
        return detalles.Select(MapToDto);
    }

    public async Task<RubroUniformeDetalleDto?> GetByIdAsync(int id)
    {
        var detalle = await _rubroUniformeDetalleRepository.GetByIdAsync(id);
        return detalle != null ? MapToDto(detalle) : null;
    }

    public async Task<IEnumerable<RubroUniformeDetalleDto>> GetByRubroIdAsync(int rubroId)
    {
        var detalles = await _rubroUniformeDetalleRepository.GetByRubroIdAsync(rubroId);
        return detalles.Select(MapToDto);
    }

    public async Task<IEnumerable<RubroUniformeDetalleDto>> GetByPrendaUniformeIdAsync(int prendaUniformeId)
    {
        var detalles = await _rubroUniformeDetalleRepository.GetByPrendaUniformeIdAsync(prendaUniformeId);
        return detalles.Select(MapToDto);
    }

    public async Task<RubroUniformeDetalleDto?> GetByRubroIdAndPrendaUniformeIdAsync(int rubroId, int prendaUniformeId)
    {
        var detalle = await _rubroUniformeDetalleRepository.GetByRubroIdAndPrendaUniformeIdAsync(rubroId, prendaUniformeId);
        return detalle != null ? MapToDto(detalle) : null;
    }

    public async Task<IEnumerable<RubroUniformeDetalleDto>> GetActiveAsync()
    {
        var detalles = await _rubroUniformeDetalleRepository.GetActiveAsync();
        return detalles.Select(MapToDto);
    }

    public async Task<RubroUniformeDetalleDto> CreateAsync(RubroUniformeDetalleCreateDto createDto, int usuarioCreacionId)
    {
        // Validate that the combination doesn't already exist
        var exists = await _rubroUniformeDetalleRepository.ExistsAsync(createDto.RubroId, createDto.PrendaUniformeId);
        if (exists)
            throw new InvalidOperationException($"A RubroUniformeDetalle already exists for RubroId {createDto.RubroId} and PrendaUniformeId {createDto.PrendaUniformeId}");

        // Validate that Rubro and PrendaUniforme exist
        var rubro = await _rubroRepository.GetByIdAsync(createDto.RubroId);
        if (rubro == null)
            throw new ArgumentException($"Rubro with ID {createDto.RubroId} not found");

        var prendaUniforme = await _prendaUniformeRepository.GetByIdAsync(createDto.PrendaUniformeId);
        if (prendaUniforme == null)
            throw new ArgumentException($"PrendaUniforme with ID {createDto.PrendaUniformeId} not found");

        var detalle = new RubroUniformeDetalle
        {
            RubroId = createDto.RubroId,
            PrendaUniformeId = createDto.PrendaUniformeId,
            FechaCreacion = DateTime.UtcNow,
            UsuarioCreacionId = usuarioCreacionId,
            EsEliminado = false
        };

        await _rubroUniformeDetalleRepository.AddAsync(detalle);

        return await GetByIdAsync(detalle.Id) ?? throw new InvalidOperationException("Failed to retrieve created detalle");
    }

    public async Task<RubroUniformeDetalleDto> UpdateAsync(int id, RubroUniformeDetalleCreateDto updateDto, int usuarioActualizacionId)
    {
        var detalle = await _rubroUniformeDetalleRepository.GetByIdAsync(id);
        if (detalle == null)
            throw new ArgumentException($"RubroUniformeDetalle with ID {id} not found");

        // Check if the new combination already exists (excluding current record)
        var existingDetalle = await _rubroUniformeDetalleRepository.GetByRubroIdAndPrendaUniformeIdAsync(updateDto.RubroId, updateDto.PrendaUniformeId);
        if (existingDetalle != null && existingDetalle.Id != id)
            throw new InvalidOperationException($"A RubroUniformeDetalle already exists for RubroId {updateDto.RubroId} and PrendaUniformeId {updateDto.PrendaUniformeId}");

        // Validate that Rubro and PrendaUniforme exist
        var rubro = await _rubroRepository.GetByIdAsync(updateDto.RubroId);
        if (rubro == null)
            throw new ArgumentException($"Rubro with ID {updateDto.RubroId} not found");

        var prendaUniforme = await _prendaUniformeRepository.GetByIdAsync(updateDto.PrendaUniformeId);
        if (prendaUniforme == null)
            throw new ArgumentException($"PrendaUniforme with ID {updateDto.PrendaUniformeId} not found");

        detalle.RubroId = updateDto.RubroId;
        detalle.PrendaUniformeId = updateDto.PrendaUniformeId;
        detalle.FechaActualizacion = DateTime.UtcNow;
        detalle.UsuarioActualizacionId = usuarioActualizacionId;

        await _rubroUniformeDetalleRepository.UpdateAsync(detalle);

        return await GetByIdAsync(id) ?? throw new InvalidOperationException("Failed to retrieve updated detalle");
    }

    public async Task<bool> DeleteAsync(int id, string motivoEliminacion, int usuarioEliminacionId)
    {
        var detalle = await _rubroUniformeDetalleRepository.GetByIdAsync(id);
        if (detalle == null)
            return false;

        detalle.EsEliminado = true;
        detalle.MotivoEliminacion = motivoEliminacion;
        detalle.FechaEliminacion = DateTime.UtcNow;
        detalle.UsuarioEliminacionId = usuarioEliminacionId;

        await _rubroUniformeDetalleRepository.UpdateAsync(detalle);
        return true;
    }

    public async Task<bool> ExistsAsync(int rubroId, int prendaUniformeId)
    {
        return await _rubroUniformeDetalleRepository.ExistsAsync(rubroId, prendaUniformeId);
    }

    private static RubroUniformeDetalleDto MapToDto(RubroUniformeDetalle detalle)
    {
        return new RubroUniformeDetalleDto
        {
            Id = detalle.Id,
            RubroId = detalle.RubroId,
            PrendaUniformeId = detalle.PrendaUniformeId,
            RubroDescripcion = detalle.Rubro?.Descripcion ?? string.Empty,
            PrendaUniformeDescripcion = detalle.PrendaUniforme?.Descripcion ?? string.Empty,
            PrendaUniformeSexo = detalle.PrendaUniforme?.Sexo ?? string.Empty,
            PrendaUniformeTalla = detalle.PrendaUniforme?.Talla ?? string.Empty,
            PrendaUniformePrecio = detalle.PrendaUniforme?.Precio ?? 0,
            PrendaUniformeImagenUrl = detalle.PrendaUniforme?.ImagenesPrenda?.FirstOrDefault()?.Imagen?.ToString(),
            PrendaUniformeExistenciaInicial = detalle.PrendaUniforme?.ExistenciaInicial ?? 0,
            PrendaUniformeEntradas = detalle.PrendaUniforme?.Entradas ?? 0,
            PrendaUniformeSalidas = detalle.PrendaUniforme?.Salidas ?? 0,
            FechaCreacion = detalle.FechaCreacion,
            FechaActualizacion = detalle.FechaActualizacion,
            UsuarioCreacionId = detalle.UsuarioCreacionId,
            UsuarioActualizacionId = detalle.UsuarioActualizacionId,
            EsEliminado = detalle.EsEliminado,
            MotivoEliminacion = detalle.MotivoEliminacion,
            FechaEliminacion = detalle.FechaEliminacion,
            UsuarioEliminacionId = detalle.UsuarioEliminacionId
        };
    }
}
